Add more specific error message when user passes incorrect matrix format to from_coo

fhoang7 opened the pull request:
- [x] closes #26554
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Commit:
Include more specific error messaging for non coo matrix and introduce non regression tests

Commit:
Log changes to what's new

jreback commented on pandas/core/arrays/sparse.py:
```diff
@@ -2014,11 +2014,15 @@ def from_coo(cls, A, dense_index=False):
+        try:
+            result = _coo_to_sparse_series(A, dense_index=dense_index,
+                                           sparse_series=False)
+            result = Series(result.array, index=result.index, copy=False)
```
you want to do this inside _coo_to_sparse

fhoang7 replied:
Done!

TomAugspurger commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -590,7 +590,7 @@ Sparse
 - Bug in :class:`SparseFrame` constructor where passing ``None`` as the data would cause ``default_fill_value`` to be ignored (:issue:`16807`)
 - Bug in :class:`SparseDataFrame` when adding a column in which the length of values does not match length of index, ``AssertionError`` is raised instead of raising ``ValueError`` (:issue:`25484`)
-
+- Introduce a more specific error message in :class:`SparseAccessor` so user knows when they provide the incorrect matrix format to `from_coo` (:issue:`26554`)
```
I don't think the ref too SparseAccessor. Probably want :meth:\`Series.sparse.from_coo\`

TomAugspurger commented on pandas/core/arrays/sparse.py:
```diff
@@ -2014,11 +2014,15 @@ def from_coo(cls, A, dense_index=False):
+            return result
+
+        except AttributeError:
+            raise AttributeError('Expected coo matrix format.')
```
Include `'Got {} instead'.format(type(A).__name__)`

TomAugspurger replied:
And this should probably be a `TypeError` rather than an AttributeError.

TomAugspurger commented on pandas/tests/arrays/sparse/test_array.py:
```diff
@@ -1128,6 +1128,13 @@ def test_from_coo(self):
         expected = pd.Series([4, 9, 7, 5], index=index, dtype='Sparse[int]')
         tm.assert_series_equal(result, expected)
 
+    @td.skip_if_no_scipy
```
Is this test different from the other one? I think we only need the one in test_accessor.py

Commit:
Remove redundant test
Move try-catch block to _coo_to_sparse_series
Update entry log

pep8speaks commented:
Hello @fhoang7! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/arrays/sparse.py`](https://github.com/pandas-dev/pandas/blob/282bc9fcfcd10b269bd7718b98b1a9197ddd6257/pandas/core/arrays/sparse.py):

> [Line 2018:44](https://github.com/pandas-dev/pandas/blob/282bc9fcfcd10b269bd7718b98b1a9197ddd6257/pandas/core/arrays/sparse.py#L2018): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

* In the file [`pandas/core/sparse/scipy_sparse.py`](https://github.com/pandas-dev/pandas/blob/282bc9fcfcd10b269bd7718b98b1a9197ddd6257/pandas/core/sparse/scipy_sparse.py):

> [Line 155:80](https://github.com/pandas-dev/pandas/blob/282bc9fcfcd10b269bd7718b98b1a9197ddd6257/pandas/core/sparse/scipy_sparse.py#L155): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 156:1](https://github.com/pandas-dev/pandas/blob/282bc9fcfcd10b269bd7718b98b1a9197ddd6257/pandas/core/sparse/scipy_sparse.py#L156): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

Commit:
Fix pep8

Commit:
Change test to expect TypeError

Commit:
Update what's new

Commit:
Update docstring to reflect what will be returned in an error state

Commit:
Merge remote-tracking branch 'upstream/master' into first-feature

fhoang7 commented:
@TomAugspurger @jreback: Changes made! Please review when you get the chance. Thanks!

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -590,7 +590,7 @@ Sparse
 - Bug in :class:`SparseFrame` constructor where passing ``None`` as the data would cause ``default_fill_value`` to be ignored (:issue:`16807`)
 - Bug in :class:`SparseDataFrame` when adding a column in which the length of values does not match length of index, ``AssertionError`` is raised instead of raising ``ValueError`` (:issue:`25484`)
-
+- Introduce a more specific error message in :meth:`_coo_to_sparse_series` so it returns a TypeError for inputs that are not coo matrices (:issue:`26554`)
```
can you use ``:meth:`Series.sparse.from_coo`` here, as the _coo...is an internal method

also say, better error message

jreback commented on pandas/core/sparse/scipy_sparse.py:
```diff
@@ -129,23 +129,29 @@ def _coo_to_sparse_series(A, dense_index: bool = False,
     -------
-    Series or SparseSeries
+    Series or SparseSeries on success
+    TypeError if A is not a coo_matrix
```
this should be in the Raises section

jreback commented on pandas/core/sparse/scipy_sparse.py:
```diff
@@ -129,23 +129,29 @@ def _coo_to_sparse_series(A, dense_index: bool = False,
+            j = range(A.shape[1])
+            ind = MultiIndex.from_product([i, j])
+            s = s.reindex(ind)
+        return s
```
son't put the try/except around everything, narrow this down to the first possible place
(which is likely
``` 
try:
    s= Series(A.data, MultiIndex.from_arrays((A.row, A.col)))
except AttributeError:
    .....
```

jreback commented on pandas/tests/arrays/sparse/test_accessor.py:
```diff
@@ -119,3 +119,10 @@ def test_series_from_coo(self, dtype, dense_index):
+
+    @td.skip_if_no_scipy
+    def test_series_from_coo_incorrect_format(self):
+        import scipy.sparse
```
add the gh issue number as a comment

Commit:
- Update what's new
- Simplify scope of except Block
- comment on non-regression test

fhoang7 commented:
@jreback: Thanks for the feedback! Changes are pushed.

jreback commented on pandas/core/sparse/scipy_sparse.py:
```diff
@@ -130,10 +130,19 @@ def _coo_to_sparse_series(A, dense_index: bool = False,
     Series or SparseSeries
+
+    Raises
+    -------
```
this is too long (has to match the length of Raises)

jreback commented on pandas/tests/arrays/sparse/test_accessor.py:
```diff
@@ -119,3 +119,11 @@ def test_series_from_coo(self, dtype, dense_index):
         tm.assert_series_equal(result, expected)
+
+    @td.skip_if_no_scipy
+    # non-regression test for fix to issue #26554
```
move the comment inside the test

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -590,7 +590,7 @@ Sparse
 - Bug in :class:`SparseFrame` constructor where passing ``None`` as the data would cause ``default_fill_value`` to be ignored (:issue:`16807`)
 - Bug in :class:`SparseDataFrame` when adding a column in which the length of values does not match length of index, ``AssertionError`` is raised instead of raising ``ValueError`` (:issue:`25484`)
-
+- Introduce a better error message in :meth:`Series.sparse.from_coo` so it returns a TypeError for inputs that are not coo matrices (:issue:`26554`)
```
use double-backtics on the TypeError

jreback reviewed (changes requested):
thank @fhoang7 looks pretty good. you will want to wait a little to merge master as turning off a doc validation check in the CI (which is causing this to fail); though there are some small issues as indicated.

Commit:
- Clean up what's new formatting
- Move comment inside test
- Update docstring delineator bar

fhoang7 commented:
Done! @jreback

simonjayhawkins commented on pandas/tests/arrays/sparse/test_accessor.py:
```diff
@@ -119,3 +119,11 @@ def test_series_from_coo(self, dtype, dense_index):
         tm.assert_series_equal(result, expected)
+
+    @td.skip_if_no_scipy
+    def test_series_from_coo_incorrect_format(self):
```
```suggestion
    def test_series_from_coo_incorrect_format_raises(self):
```

fhoang7 replied:
I like that. added type error to the end to be more specific.

simonjayhawkins replied:
having _raises at the end of the test name is for consistency. if you can lose the type error bit that'll be great

fhoang7 replied:
will do!

simonjayhawkins commented on pandas/tests/arrays/sparse/test_accessor.py:
```diff
@@ -119,3 +119,11 @@ def test_series_from_coo(self, dtype, dense_index):
+        # non-regression test for fix to issue #26554
+        import scipy.sparse
+        m = scipy.sparse.csr_matrix(np.array([[0, 1], [0, 0]]))
+        with pytest.raises(TypeError):
```
can you add a message check

fhoang7 replied:
Done!

simonjayhawkins replied:
as this test is not parametrised can you just use the literal message raised by this test instead.

fhoang7 replied:
Will do!

Commit:
Add message assertion to test and change test name to be more verbose in functionality

fhoang7 commented:
@simonjayhawkins: Made requested changes. Thanks for your feedback!

simonjayhawkins commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -590,7 +590,7 @@ Sparse
 - Bug in :class:`SparseFrame` constructor where passing ``None`` as the data would cause ``default_fill_value`` to be ignored (:issue:`16807`)
 - Bug in :class:`SparseDataFrame` when adding a column in which the length of values does not match length of index, ``AssertionError`` is raised instead of raising ``ValueError`` (:issue:`25484`)
-
+- Introduce a better error message in :meth:`Series.sparse.from_coo` so it returns a `TypeError` for inputs that are not coo matrices (:issue:`26554`)
```
should be double ticks

simonjayhawkins commented on pandas/tests/arrays/sparse/test_accessor.py:
```diff
@@ -119,3 +119,13 @@ def test_series_from_coo(self, dtype, dense_index):
+
+    @td.skip_if_no_scipy
+    def test_series_from_coo_incorrect_format_raises_type_error(self):
+        # non-regression test for fix to issue #26554
```
```suggestion
        # gh-26554
```

Commit:
Proper formatting in what's new + proper commenting for test

fhoang7 commented:
@simonjayhawkins: Pushed suggested changes.

simonjayhawkins commented:
lgtm. not sure if you saw the comments on the error mesage check and the test name. if you can address those i think we're done

fhoang7 commented:
@simonjayhawkins: Yup latest commit reflect those changes. Only thing is not sure why some checks are fixing so I only changed comments/test name.

simonjayhawkins commented:
> Only thing is not sure why some checks are fixing so I only changed comments/test name. Looks like similar errors in the past builds before me.

failing CI is not related to your changes

> Yup latest commit reflect those changes

please address https://github.com/pandas-dev/pandas/pull/26584#discussion_r289621003 and https://github.com/pandas-dev/pandas/pull/26584#discussion_r289621024

Commit:
Fix test name for consistency and assert on literal message

fhoang7 commented:
@simonjayhawkins: Thanks for your patience and input. Addressed those comments and double checked over your previous comments as well to make sure I resolved them.

simonjayhawkins reviewed (approved):
thanks @fhoang7 approved assuming green on py3.5 @TomAugspurger @jreback

simonjayhawkins reviewed (changes requested):
@fhoang7 failing CI is now fixed. can you merge master.

Commit:
Merge remote-tracking branch 'upstream/master' into first-feature

fhoang7 commented:
@simonjayhawkins: Thanks for the headsup. Merged/pushed up.

jreback commented:
thanks @fhoang7
